eos-updater/poll.c

```diff
@@ -494,6 +498,22 @@ check_for_update_following_checkpoint_commits (OstreeRepo     *repo,
                               error))
     return FALSE;
 
+  /* FIXME: Once we’ve worked out which commit to update to, we really should
+   * then iterate back through the commits on @new_refspec until we get to:
```

Comment:
I'm a bit sceptical of the idea that determining whether updating from A to B should be treated as a major upgrade rather than a minor update should involve iterating over the complete commit history:

1. The server doesn't necessarily have the complete history for a ref. I checked on the server and the `os/eos/amd64/eos3a` only has versions `3.9.0` through `3.9.7`. I think this is an artefact of how our branching works, basically the `os/eos/amd64/eos3a` is an alias for the latest 3.x branch, `os/eos/amd64/eos3.9`. In the dev repo, `os/eos/amd64/master` only goes back 5 days. (For releases we do have the commits so we could in principle reconstruct this history. For nightlies we don't keep old commits.)
2. Each `.commit` object is like 4.3MB. If you have to fetch dozens of these it could end up comparable to the size of the update itself. And you either have to store them, or re-fetch them on each poll (in the case where the update is not fetched).
3. The client may not be able to fetch the history. Currently, if you have a random OS image it can be used as a USB updates source without further work. But our OS images don't include the complete ostree ref history for the commit within.

Here's another idea. Our version numbers take the form X.Y.Z, where X changes when we want it to; Y changes when we rebranch the OS (which we might do for reasons uninteresting to the typical user, such as branching 5.1 soon after 5.0 to introduce a newer kernel); and Z changes on point releases. Can we simply declare that we will give any new OS release that warrants a big banner a new X, as we did with EOS 4 (removed a bunch of features) and will do with EOS 5? Then we could determine whether updating from A to B is a major upgrade or minor update by comparing the version numbers, which is already a property of both commits. This would be consistent with how GNOME is versioned; and how (say) Fedora treats upgrades from a user's perspective, where 32 → 33 is noteworthy but miscellaneous package updates within a series is not.

Reply:
Those are all good arguments against this design. I was shying away from comparing version numbers (or branch names) because of the fear of overloading their semantics, but I guess the version number semantics are an exact match for this, on EOS at least.

It won’t necessarily be such a great match for other distros who use/want to use eos-updater, but perhaps let’s cross that bridge when we come to it.

Comparing version numbers also isn’t a great match for the other potential uses I had in mind for `UpdateFlags` — indicating update urgency and severity. But since nobody’s asked for those features, they shouldn’t prejudice the implementation of this one. We can always implement them a different way in future, if anyone ever asks for them.

I’m looking at reworking this to compare version numbers. The `UpdateFlags` D-Bus property can stay (although perhaps I’ll change it to a stand-alone `UpdateIsUserVisible` property) but the `eos-updater.update-flags` commit metadata key will be removed. That means the corresponding [eos-build PR](https://github.com/endlessm/eos-build/pull/1398) will probably get closed unmerged.

Reply:
Significantly reworked it to compare version numbers instead.

Reply:
Definitely for EOS the release versions follow semantic versioning. I may be wrong, but I think I even added some code in the release tooling that ensures the version isn't a downgrade. So, I'd say for EOS comparing version numbers should work fine. Certainly there's code in eos-build/ostree that decides on major vs minor upgrade based entirely on version numbers.

(There's a bit of minutiae about how pre-releases are handled. The commit metadata version field uses `~` like in Debian version handling, but `~` isn't allowed in OSTree refs, so it's converted to `-` for the `release/*` refs.)